backend/ee/danswer/server/tenants/api.py

```diff
@@ -60,3 +74,53 @@ def create_tenant(
     finally:
         if token is not None:
             current_tenant_id.reset(token)
+
+
+@router.post("/product-gating")
+def gate_product(
+    product_gating_request: ProductGatingRequest, _: None = Depends(control_plane_dep)
+) -> None:
+    """
+    Gating the product means that the product is not available to the tenant.
+    They will be directed to the billing page.
+    We gate the product when
+    1) User has ended free trial without adding payment method
+    2) User's card has declined
```

Comment:
We can let them use it and give a warning if the card declined. Hubspot does this and I appreciated that and paid them :P